### PR TITLE
Add an audit for mismatched Python resource and PyPi package names

### LINKED ISSUE #14537: audit: flag Python/PyPI resources with mismatching names/PyPI package names
### Verification

- [X] This issue's title and/or description do not reference a single formula e.g. `brew install wget`. If they do, open an issue at https://github.com/Homebrew/homebrew-core/issues/new/choose instead.

### Provide a detailed description of the proposed feature

This is a nice-to-have `audit` feature: when people manually add or modify PyPI resources, the `Resource#name` sometimes ends up out-of-sync with the PyPI package name.

For example:

```ruby
  resource "Purchasing" do
    url "https://files.pythonhosted.org/packages/41/2a/354f0ae3fa02708719e2ed6a8c310da4283bf9a589e2a7fcf7dadb9638af/PyQtPurchasing-5.15.5.tar.gz"
    sha256 "8bb1df553ba6a615f8ec3d9b9c5270db3e15e831a6161773dabfdc1a7afe4834"
  end
```

should be:

```ruby
  resource "PyQtPurchasing" do
    url "https://files.pythonhosted.org/packages/41/2a/354f0ae3fa02708719e2ed6a8c310da4283bf9a589e2a7fcf7dadb9638af/PyQtPurchasing-5.15.5.tar.gz"
    sha256 "8bb1df553ba6a615f8ec3d9b9c5270db3e15e831a6161773dabfdc1a7afe4834"
  end
```

I fixed up a few of these manually (e.g. in `pyqt`), but detecting them automatically would be nice!

I can take a stab at this in a bit if nobody else is interested; just recording it publicly with this issue.

### What is the motivation for the feature?

Quality of life/accuracy of `resource` blocks; this also ensures that some of our automation (`brew-pip-audit`) continues to function reliably.

### How will the feature be relevant to at least 90% of Homebrew users?

No direct effect on end users; primarily relevant for overall codebase quality management and automation.

### What alternatives to the feature have been considered?

Continuing to do it manually.

### PR DESCRIPTION
**Pull request by issyl0**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

- Fixes #14537.
- When people manually add or modify PyPI resources the `Resource#name` sometimes ends up out-of-sync with the PyPI package name.

### CODE REVIEW

File: Library/Homebrew/resource_auditor.rb [hunk 1]
```diff
@@ -100,6 +100,15 @@ def self.curl_deps
       end
     end
 
+    def audit_resource_name_matches_pypi_package_name_in_url
+      return unless url.match?(%r{^https?://files\.pythonhosted\.org/packages/})
+
+      pypi_package_name = url.split("/").last.split(/-\d+\.\d+./).first.tr("_", "-")
```
Comment (Bo98) [inline Library/Homebrew/resource_auditor.rb:106]: Here's other characters of interest: https://github.com/Homebrew/brew/blob/e7d0625776dca58c81b03c19eeda1af8bd6bb58e/Library/Homebrew/utils/pypi.rb#L345-L349

File: Library/Homebrew/resource_auditor.rb [hunk 2]
```diff
@@ -100,6 +100,15 @@ def self.curl_deps
...
+      return if name.casecmp(pypi_package_name).zero?
+
+      problem "resource name should be `#{pypi_package_name}` to closer match the PyPI package name"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/resource_auditor.rb:109]:
```suggestion
      problem "resource name should be `#{pypi_package_name}` to match the PyPI package name"
```

maybe?

**Review by MikeMcQuaid — COMMENTED** [review]
Great idea!

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by issyl0** [comment]
🤔 Are these findings correct?

```
alot
  * Stable resource "zope.interface": resource name should be `zope-interface` to match the PyPI package name
  * HEAD resource "zope.interface": resource name should be `zope-interface` to match the PyPI package name
```

The PyPI package name is `zope.interface` so the resource name is `zope.interface` and that's legit. But based on the other normalization code we have, I changed it to translate `.` and `_` into `-`. But presumably keeping the dots in the resource name was actually intentional? What should we do here?

**Comment by Bo98** [comment]
I don't think there's any particular reason except that `brew update-python-resources` hasn't touched it

If you want, you can relax it like how Python behaves: https://packaging.python.org/en/latest/specifications/name-normalization/ (i.e. also apply the `[-_.]+` gsub to the resource name). We probably don't need to be picky here, especially if we apply this to third-party taps.

**Comment by issyl0** [comment]
Yeah I just re-read the normalization docs for PyPI package names, it's all good, they should change (not that it really matters - we support -_. but hey).

File: Library/Homebrew/resource_auditor.rb [hunk 3]
```diff
@@ -100,6 +100,16 @@ def self.curl_deps
       end
     end
 
+    def audit_resource_name_matches_pypi_package_name_in_url
+      return unless url.match?(%r{^https?://files\.pythonhosted\.org/packages/})
+      return if name == owner.name # Skip the top-level package name as we only care about `resource "foo"` blocks.
+
+      pypi_package_name = url.split("/").last.split(/[-.]\d+?./).first.gsub(/[_.]/, "-")
```
Comment (MikeMcQuaid) [inline Library/Homebrew/resource_auditor.rb:107]: A regex might be easier to follow than the `split`s?

**Review by MikeMcQuaid — APPROVED** [review]
Looks good! One question but happy to ship as-is. ✅ 🚢 once homebrew/core audit failure is resolved.

**Comment by MikeMcQuaid** [comment]
Thanks again @issyl0!

**Pull request merged**